amaru/src/mempool/mod.rs

```diff
@@ -0,0 +1,104 @@
+use crate::ledger;
+use pallas_primitives::conway::Tx;
+use std::collections::BTreeSet;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+// State
+// ----------------------------------------------------------------------------
+
+/// The state of the mempool is split into two sub-components:
+///
+/// - A _ledger_state_ which contains a view of a ledger state, which is used to validate
+///   transactions.
+///
+/// - An _ephemeral_ state, which is maintained as a sequence of diff operations to be applied on
+///   top of the _ledger_state_. It should be re-calculated as transactions are added or removed
+///   from the mempool.
+pub struct State<S, E>
+where
+    S: ledger::store::Store<Error = E>,
+{
+    /// The current view of the ledger state.
+    ledger_state: ledger::state::State<S, E>,
+
+    /// The ephemeral state, which is applied on top of the ledger_state using transactions
+    /// in the mempool.
+    ephemeral: ledger::state::VolatileDB,
```

Comment:
The ephemeral state is where we apply the mempool txns on top of the existing ledger state. This is the same object that is used internally in the ledger state to keep part of it volatile and subject to rollbacks. I feel like the same thing could be utilized in a mempool.  So, a LedgerState has both stable and volatile transactions.  A MempoolState has a stable, volatile, and ephemeral transactions.

So, let's say a rollback occurs on-chain in the volatile section, that should trigger a rebuilding of the ephemeral transactions.

It could also be that a transaction is removed from the mempool by a user, or something is re-ordered. This should also trigger a rebuild of the ephemeral transactions.

When we go to make a block, we have a nice ordered list of transactions in the ephemeral state we can take a bite of and submit as a block.

Reply:
My understanding of the purpose of this structure from my discussions with @KtorZ is that it's sequential and meant for fast and easy handling of rollbacks: Should you have a rollback to some previous transaction, you  retrieve the cached state from the queue then reapply from that point on. This is useful in the context of the ledger/consensus because there might be a large number of blocks and transactions involved.

My understanding is that the behaviour of the mempool is different: It's perfectly possible that given a list of transactions say. `[tx1, tx2 ... tx_10]` in the mempool, you would need to remove say `tx_5` and `tx_8`, which would mean that you would rollback to `tx_4` and then need to reapply the remaining txs. Given the expected size of the mempool, and the fact you don't need to redo phase 2 validations but only phase 1 which is very fast, I would argue that this `VolatileState` is a premature optimisation.

Reply:
Indeed; I am not sure the volatile state is best suited for storing the transient mempool. Mostly also because it's a VecDeque, which has fast access to both ends of the vector but not much to arbitrary elements. 

For a mempool, we might need something with fast access to arbitrary transactions and possibly with extra metadata such as their size and execution units. 


To be frank, I find the weighted RoseTree from the Haskell implementation not bad; with good amortised performances. Another option could be a weighted Patricia trie; which should work quite well with transaction ids as keys.

Reply:
Given we don't really know, this is a cache and optimisation, and this might be heavily dependent on performance measures, I would leave that out for the moment and go for the simplest solution: 
* Keep 2 states an "anchor" and a "tip"
* Every time txs are added, update the "tip"
* Every time txs are removed, revalidate all txs and define new "tip" 
* Every time underlying chain changes, change "anchor", revalidate txs and define new "tip"